### PR TITLE
Real Double Class

### PR DESCRIPTION
**Pull request by rajithv**
I have several points to clear up

I used sympify rather than initialize (same as in Complex and Rational classes). But with this the Integer class seems to be the only one different from the rest. Is it okay if we go ahead and change the integer class to be constructed throught sympify. 
i.e. rather than `A = SymEngine::Integer.new(5)` we can use `A = SymEngine(5)`.
Also, the macro `GET_SYMINTFROMVAL(num_value, this)` can be incorporated into `sympify` method in `symengine_utils.c`.

Please advice how to proceed. @abinashmeher999  @isuruf 

P.S. Also, all the commits from the other PR came into this as well.. will this be an issue ?

### CODE REVIEW

**Comment by rajithv** [comment]
Another uncertainity, as per the following code:

```
2.1.8 :005 > c4 = 4.5 + (7i/9)
 => (4.5+(7/9)*i) 
2.1.8 :006 > c5 = SymEngine(c4)
 => #<SymEngine::ComplexDouble(4.5 + 0.777777777777778*I)>
```

What should we do with this kind of complex numbers?

**Comment by isuruf** [comment]
That's fine. It was a design decision of symengine to keep it like that. Here, real part is machine precision, while imaginary part is infinite precision. Most often you don't want to have the real part in one precision and the imaginary part in another.

**Comment by isuruf** [comment]
Can you merge with master?

File: ext/symengine/symengine_utils.c [hunk 1]
```diff
@@ -6,13 +6,23 @@ void sympify(VALUE operand2, basic_struct *cbasic_operand2) {
...
     switch(TYPE(operand2)) {
         case T_FIXNUM:
         case T_BIGNUM:
             GET_SYMINTFROMVAL(operand2, cbasic_operand2);
             break;
 
+        case T_FLOAT:         
+            Rb_Temp_String = rb_funcall(operand2, rb_intern("to_s"), 0, NULL);           
+
+            s = StringValueCStr(Rb_Temp_String);
```
Comment (isuruf) [inline ext/symengine/symengine_utils.c]: Use `rb_float_value`. Don't use strings unless you really have to.

**Comment by abinashmeher999** [comment]
@rajithv You should first understand why were the initialize methods removed for `Rational` and `Complex` classes. Integers are a subset of Rationals, Rationals of Real and Real of Complex. So when a complex is initialised with imaginary part as 0, it should fall down to Real or Rational. Likewise when a Rational is initialised with 1 as denominator it should fall down to Integer. So it made no sense to output an Integer when `.new` was being called for a Rational number. But is it so with Integer?

For `SymEngine(5)`, it should still give an Integer. So to have this, `.new` for Integer need not be removed.

I don't see the reason to remove `GET_SYMINTFROMVAL` if that is what you mean by saying "incorporating it to `sympify`" Someone would use that macro when (s)he is expecting an Integer from VALUE and nothing else, so that if a wrong input comes an exception is thrown.

File: ext/symengine/symengine_utils.c (hunk 1 above)
Comment (abinashmeher999) [inline ext/symengine/symengine_utils.c]: Yes, that will be much faster than this.

File: lib/symengine/real_double.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module SymEngine
+  class RealDouble
+    def to_f
+        self.to_s.to_f
```
Comment (abinashmeher999) [inline lib/symengine/real_double.rb]: Same here. You can do this in C API without having to convert it into string first.

**Comment by rajithv** [comment]
@abinashmeher999 Thanks for the explanation.

I meant to remove the `GET_SYMINTFROMVAL` macro, and have that code inside `sympify` under the 
```
case T_FIXNUM:
case T_BIGNUM:
```
then someone expecting a Integer from a VALUE can simply use simpify, with and if they expect nothing else use an assertion on that side. Or we can have a method, which calls sympify and raises an exception if sympify returns a non-SymEngine::Integer. This is my suggestion. What do you think?

**Comment by isuruf** [comment]
I think the current way is fine. You can change `GET_SYMINTFROMVAL` macro to a function. See #47

**Comment by abinashmeher999** [comment]
That can also be done. Think about the cases when the user doesn't want anything other than Integer. You do some computation to `sympify` it to some other Class only to be rejected later, and that much computation is wasted.

If you could somehow bring the selection part to before the computation is done to convert it then it would be great.

**Comment by rajithv** [comment]
The `rb_float_value` failes on Ruby 1.9.3 .. it is not in ruby versions prior to 2.

see : [Ruby 1.9.3](http://rxr.whitequark.org/mri/ident?v=2.0.0-p481;i=rb_float_value) vs [Ruby 2.0.0] (http://rxr.whitequark.org/mri/ident?v=1.9.3-p547;i=rb_float_value)

Is there a way to do version dependent coding for the Ruby C API?

**Comment by isuruf** [comment]
Use `RFLOAT_VALUE`

**Comment by v0dro** [comment]
Looks good. What will be the next class you plan to expose through Ruby?

**Comment by zverok** [comment]
(BTW, I'm not sure somebody should support 1.9.3 in a new Ruby gems in 2016, as even Ruby 2.0 is [not officially supported](https://www.ruby-lang.org/en/news/2016/02/24/support-plan-of-ruby-2-0-0-and-2-1/) anymore.)

**Pull request merged**